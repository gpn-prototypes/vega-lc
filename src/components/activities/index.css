.ActivitiesWidget {
  position: absolute;

  margin: 16px;
  padding: 8px;

  width: 40%;
  max-width: 316px;
  height: 840px;

  background-color: var(--color-bg-default);

  &__Title {
    margin: 0;

    font-size: 16px;
    font-style: normal;
    line-height: 150%;
  }

  & .VegaTree {
    padding: 0;
  }
}
