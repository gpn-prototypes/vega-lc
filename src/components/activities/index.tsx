import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IconSearch, TargetData, Text, TextField, Tree } from '@gpn-prototypes/vega-ui';

import {
  setActivitiesDraggingElements,
  setActivitiesRef,
  setSearchString,
} from '../../redux-store/activities/actions';
import {
  getActivitiesNodeList,
  getIsAutoFocus,
  getSearchStringValue,
} from '../../redux-store/activities/selectors';

import { cnActivities } from './cn-activities';
import { BlueLineSvg } from './icons';

import './index.css';

const icons = {
  'blue-line': BlueLineSvg,
};

export const ActivitiesWidget: React.FC = (): React.ReactElement => {
  const dispatch = useDispatch();

  const activitiesRef = useRef(null);
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    dispatch(setActivitiesRef(activitiesRef));
  }, [dispatch, activitiesRef]);

  const searchString = useSelector(getSearchStringValue);
  const activities = useSelector(getActivitiesNodeList(searchString));
  const autoFocus = useSelector(getIsAutoFocus);

  const handleSearch = (args: { value: string | null }): void => {
    dispatch(setSearchString(args.value));
  };

  const handleDragStart = (transferringElems: TargetData[]): void => {
    dispatch(setActivitiesDraggingElements(transferringElems));
  };

  const handleDragEnd = (): void => {
    dispatch(setActivitiesDraggingElements([]));
  };

  return (
    <div ref={activitiesRef} className={cnActivities()}>
      <div className={cnActivities('Head')}>
        <Text className={cnActivities('Title').toString()}>Мероприятия</Text>

        <TextField
          autoFocus={autoFocus}
          className={cnActivities('Search').toString()}
          leftSide={IconSearch}
          size="s"
          inputRef={inputRef}
          type="input"
          onChange={handleSearch}
          value={searchString}
          placeholder="Поиск"
        />
      </div>

      <Tree
        icons={icons}
        nodeList={activities || []}
        withVisibilitySwitcher={false}
        withDropZoneIndicator={false}
        showIndentGuides={false}
        onDragStart={handleDragStart}
        onDragEnd={handleDragEnd}
      />
    </div>
  );
};
