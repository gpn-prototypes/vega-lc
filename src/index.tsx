import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { ApolloProvider } from '@apollo/client';
import { Root } from '@gpn-prototypes/vega-root';

import getStore from './redux-store/index';
import { App } from './App';
import client from './client';

import './App.css';

ReactDOM.render(
  <Root initialPortals={[{ name: 'modalRoot' }]} defaultTheme="dark">
    <ApolloProvider client={client}>
      <Provider store={getStore()}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    </ApolloProvider>
  </Root>,
  document.getElementById('root'),
);
