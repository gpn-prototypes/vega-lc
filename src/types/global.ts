import ResizeObserver from 'resize-observer-polyfill';

declare global {
  interface Window {
    ResizeObserver: typeof ResizeObserver;
  }
}
