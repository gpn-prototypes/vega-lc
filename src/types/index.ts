export interface Identity {
  getToken(): Promise<string>;
}
