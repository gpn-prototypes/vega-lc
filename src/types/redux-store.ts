import React from 'react';
import { CanvasData, CanvasTree, TargetData, TreeItem } from '@gpn-prototypes/vega-ui';
import { CanvasView } from '@gpn-prototypes/vega-ui/dist/components/canvas/entities/CanvasView';

type NodeList = TreeItem[];

export declare type Content = {
  type: string;
  id: string;
  name: string;
};
export declare type Event = {
  id: string;
  name: string;
  content: Content[];
};
export declare type StepData = {
  id: string;
  name: string;
  events: Event[];
};

export type SplitPanesGridParams = {
  width: string;
  height?: string;
};

export type CanvasViewEntity = CanvasView;

export type StepContent = {
  activity?: {
    activityType: {
      vid: string;
    };
    name: string;
  };
  object?: {
    vid: string;
    name: string;
  };
};

export type Step = {
  name: string;
  vid: string;
  itemList: StepContent[];
};

export type CanvasElement = {
  id?: string;
  parentIds?: string[];
  childrenIds?: string[];
  data: CanvasData;
};

export type CanvasElements = {
  vid: string;
  children?: { vid: string }[];
  parents?: { vid: string }[];
  position: [number, number];
  nodeRef?: string | null;
  nodeType: 'root' | 'end' | 'domainObject';
  title?: string;
  width?: number;
};

export type ActivitiesState = {
  nodeList: NodeList;
  searchString: string | null;
  autoFocus: boolean;
  activitiesRef?: React.RefObject<HTMLElement> | null;
  draggingElements: TargetData[] | [];
  isDroppingOnExistingStep?: boolean;
  isActivitiesPanelOpen: boolean;
};

export type LogicConstructorState = {
  scenarioList?: Step[];
  canvasElements?: CanvasTree[];
  isStepEditorOpened?: boolean;
  canvasViewRef?: React.MutableRefObject<CanvasViewEntity | null>;
};

export type NewGroupParams = {
  isDynamic: boolean;
  name: string;
};

export type GroupObjectsState = {
  nodeList: NodeList;
  isDialogOpened: boolean;
  newGroupParams: NewGroupParams;
  draggingElements: TargetData[] | [];
};

export interface ProjectStructureQuery {
  query: string;
  tree: string[];
}

export type ProjectStructureState = {
  nodeList: NodeList;
  draggingElements: TargetData[] | [];
  projectStructureQuery?: ProjectStructureQuery;
};

export type VersionState = {
  version?: number;
};

export type StoreLC = {
  activities: ActivitiesState;
  groupObjects: GroupObjectsState;
  projectStructure: ProjectStructureState;
  logicConstructor: LogicConstructorState;
  version: VersionState;
  notifications: NotificationState;
};

export type NotificationStatus = 'alert' | 'normal' | 'system' | 'success' | 'warning' | undefined;

export type Notification = {
  message: string;
  status?: NotificationStatus;
  autoClose?: boolean;
  icon?: React.FC<any>;
};

export type NotificationState = {
  list: Notification[];
};
