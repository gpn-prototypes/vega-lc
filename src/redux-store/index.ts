import { applyMiddleware, combineReducers, createStore, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore } from 'redux-persist';
import thunkMiddleware from 'redux-thunk';

import activitiesReducer from './activities/reducer';
import groupObjectsReducer from './group-objects/reducer';
import logicConstructorReducer from './logic-constructor/reducer';
import notificationsReducer from './notifications/reducer';
import projectStructureReducer from './project-structure/reducer';
import versionReducer from './version/reducer';

import { StoreLC } from '@/types/redux-store';

const rootReducer = combineReducers({
  activities: activitiesReducer,
  groupObjects: groupObjectsReducer,
  projectStructure: projectStructureReducer,
  logicConstructor: logicConstructorReducer,
  version: versionReducer,
  notifications: notificationsReducer,
});

const getStore = (initialState?: StoreLC): Store =>
  createStore(
    rootReducer,
    initialState,
    composeWithDevTools(applyMiddleware(thunkMiddleware.withExtraArgument({}))),
  );

export const store = getStore();

export const persistor = persistStore(store);
