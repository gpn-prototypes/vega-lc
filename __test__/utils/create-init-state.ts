import ActivitiesInitState from '../../src/redux-store/activities/initial-state';
import GroupObjectsInitState from '../../src/redux-store/group-objects/initial-state';
import LogicConstructorInitState from '../../src/redux-store/logic-constructor/initial-state';
import NotificationInitState from '../../src/redux-store/notifications/initial-state';
import ProjectStructureInitState from '../../src/redux-store/project-structure/initial-state';
import VersionInitState from '../../src/redux-store/version/initial-state';
import { StoreLC } from '../../src/types/redux-store';

export const createInitState = (modifier?: Partial<StoreLC>): StoreLC => {
  let initState: StoreLC = {
    activities: ActivitiesInitState,
    groupObjects: GroupObjectsInitState,
    projectStructure: ProjectStructureInitState,
    logicConstructor: LogicConstructorInitState,
    version: VersionInitState,
    notifications: NotificationInitState,
  };

  if (modifier) {
    initState = {
      ...initState,
      ...modifier,
    };
  }

  return initState;
};
