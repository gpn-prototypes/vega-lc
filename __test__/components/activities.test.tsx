import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import { Store } from 'redux';
import ResizeObserver from 'resize-observer-polyfill';

import '../../src/types/global';

import { ActivitiesWidget } from '../../src/components/activities';
import { getStore } from '../../src/redux-store';
import { StoreLC } from '../../src/types/redux-store';
import { createInitState } from '../utils/create-init-state';

let reduxStore: Store | null;

const renderActivities = (initialState?: StoreLC) => {
  const store = getStore(initialState);
  reduxStore = store;

  return render(
    <Provider store={store}>
      <ActivitiesWidget />
    </Provider>,
  );
};

beforeAll(() => {
  window.ResizeObserver = ResizeObserver;
});

afterEach(() => {
  reduxStore = null;
});

describe('Список мероприятий', () => {
  test('рендерится без ошибок', () => {
    renderActivities();
  });

  test('Поисковая строка принимает и выводит введенное значение', () => {
    const { container } = renderActivities();

    const search = container.getElementsByTagName('input')[0];

    if (search) {
      const searchString = 'Геодезия';

      expect(search.value).toBe('');

      fireEvent.change(search, { target: { value: searchString } });

      expect(search.value).toBe(searchString);
    }
  });

  test('Мероприятия можно скрыть и раскрыть', () => {
    const { container } = renderActivities();

    const button = container.querySelector('.ActivitiesWidget__Switch');
    const activities = container.querySelector('.ActivitiesWidget');

    if (button instanceof HTMLElement && activities instanceof HTMLElement) {
      expect(activities).toBeInTheDocument();
      expect(button).toBeInTheDocument();

      expect(activities.hidden).toBe(false);

      fireEvent.click(button);

      expect(activities.hidden).toBe(true);

      fireEvent.click(button);

      expect(activities.hidden).toBe(false);
    }
  });

  test('Перетягиваемый элемент корректно проставляется при событии drag-start и удаляется при drag-end', () => {
    const initialState = createInitState({
      activities: {
        isActivitiesPanelOpen: true,
        nodeList: [
          {
            name: 'mock',
            id: '1',
            nodeList: [
              {
                name: 'mock-2',
                id: '2',
                nodeList: [],
              },
            ],
          },
        ],
      },
    });

    const { container } = renderActivities(initialState);

    const element = container.querySelector('.VegaTree__TreeNode');

    if (element) {
      fireEvent.dragStart(element);

      expect(reduxStore?.getState().activities.draggingElements).toHaveLength(1);

      fireEvent.dragEnd(element);

      expect(reduxStore?.getState().activities.draggingElements).toHaveLength(0);
    }
  });
});
