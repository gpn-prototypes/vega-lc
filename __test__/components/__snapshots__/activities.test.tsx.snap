// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geological Exploration рендерится без ошибок 1`] = `
<div
  className="ActivitiesWidget"
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <div
      className="Text Text_size_m Text_view_primary ActivitiesWidget__Title"
    >
      Мероприятия
    </div>
    <div
      className="TextField TextField_size_s TextField_view_default TextField_form_default TextField_width_default TextField_type_input ActivitiesWidget__Search"
    >
      <div
        className="TextField-Side TextField-Side_position_left TextField-Side_type_icon"
      >
        <span
          className="Icon Icon_size_s IconSearch TextField-Icon"
        >
          <svg
            className="Icon-Svg"
            viewBox="0 0 16 16"
          >
            <path
              clipRule="evenodd"
              d="M9.618 11.032a5.5 5.5 0 111.414-1.414l3.97 3.969L13.586 15l-3.97-3.97zM10 6.5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0z"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </div>
      <input
        autoFocus={false}
        className="TextField-Input"
        id=""
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Поиск"
        type="input"
        value=""
      />
    </div>
  </div>
  <div
    className="VegaTree"
  >
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <ul
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        role="tree"
        tabIndex={0}
      />
    </div>
  </div>
</div>
`;
